### PR TITLE
Use abstract classes

### PR DESCRIPTION
**Pull request by naoyak**
Ref: #1130

Cherry-picking some commits from #1562.

### CODE REVIEW

File: nltk/ccg/api.py [hunk 1]
```diff
@@ -5,43 +5,58 @@
...
+         - Returns None if they cannot be unified
+         - Returns a list of necessary substitutions if they can.
+        """
 
     # Utility functions: comparison, strings and hashing.
-
+    @abstractmethod
     def __str__(self):
-        raise NotImplementedError()
+        pass
```
Comment (alvations) [inline nltk/ccg/api.py:60]: The `pass` here is inconsistent with the rest of the `@abstractmethod`. Is it necessary?

Comment (alvations) [reply nltk/ccg/api.py:60]: Ah my mistake there's no docstring for this, that's why =)

File: nltk/metrics/association.py [hunk 2]
```diff
@@ -61,16 +64,14 @@ class NgramAssocMeasures(object):
     _n = 0
 
     @staticmethod
+    @abstractmethod
     def _contingency(*marginals):
         """Calculates values of a contingency table from marginal values."""
-        raise NotImplementedError("The contingency table is not available"
-                                    "in the general ngram case")
```
Comment (alvations) [inline nltk/metrics/association.py]: For these cases I think the error message is important. Is there a way to use `@abstractmethod` and still raise the error message?

File: nltk/metrics/association.py [hunk 3]
```diff
@@ -61,16 +64,14 @@ class NgramAssocMeasures(object):
...
 
     @staticmethod
+    @abstractmethod
     def _marginals(*contingency):
         """Calculates values of contingency table marginals from its values."""
-        raise NotImplementedError("The contingency table is not available"
-                                    "in the general ngram case")
```
Comment (alvations) [inline nltk/metrics/association.py]: Same as above. 

For these cases I think the error message is important. Is there a way to use `@abstractmethod` and still raise the error message?

File: nltk/metrics/association.py (hunk 2 above)
Comment (naoyak) [reply nltk/metrics/association.py]: good idea, added it back!

File: nltk/metrics/association.py (hunk 3 above)
Comment (naoyak) [reply nltk/metrics/association.py]: good idea, added it back!

**Comment by naoyak** [comment]
@alvations thanks for the feedback - reflected

**Comment by naoyak** [comment]
Rebased and updated.

**Comment by naoyak** [comment]
@alvations @stevenbird - does this look better?

**Comment by stevenbird** [comment]
@naoyak – yes, looking good!

**Pull request merged**